# **Polars**

In [ ]:
import polars as pl

In [ ]:
# Загрузка датасета с помощью polars
df = pl.read_csv("train.csv")

In [ ]:
# Выводим информацию о типах данных и количестве пропусков
print("Типы данных:")
print(df.dtypes)

print("\nКоличество пропусков:")
print(df.null_count())

print("\nОсновные статистики:")
print(df.describe())

print("\nСредние значения числовых столбцов:")
print(df.mean())


Типы данных:
[Int64, Int64, Int64, String, String, Float64, Int64, Int64, String, Float64, String, String]

Количество пропусков:
shape: (1, 12)
┌─────────────┬──────────┬────────┬──────┬───┬────────┬──────┬───────┬──────────┐
│ PassengerId ┆ Survived ┆ Pclass ┆ Name ┆ … ┆ Ticket ┆ Fare ┆ Cabin ┆ Embarked │
│ ---         ┆ ---      ┆ ---    ┆ ---  ┆   ┆ ---    ┆ ---  ┆ ---   ┆ ---      │
│ u32         ┆ u32      ┆ u32    ┆ u32  ┆   ┆ u32    ┆ u32  ┆ u32   ┆ u32      │
╞═════════════╪══════════╪════════╪══════╪═══╪════════╪══════╪═══════╪══════════╡
│ 0           ┆ 0        ┆ 0      ┆ 0    ┆ … ┆ 0      ┆ 0    ┆ 687   ┆ 2        │
└─────────────┴──────────┴────────┴──────┴───┴────────┴──────┴───────┴──────────┘

Основные статистики:
shape: (9, 13)
┌────────────┬─────────────┬──────────┬──────────┬───┬───────────┬───────────┬───────┬──────────┐
│ statistic  ┆ PassengerId ┆ Survived ┆ Pclass   ┆ … ┆ Ticket    ┆ Fare      ┆ Cabin ┆ Embarked │
│ ---        ┆ ---         ┆ ---      ┆ ---     

In [ ]:
# Просчет количества пассажиров каждого класса (Pclass)
pclass_counts = df.get_column('Pclass').value_counts().sort(by='Pclass')
pclass_counts

Pclass,count
i64,u32
1,216
2,184
3,491


In [ ]:
# Количество выживших мужчин и женщин на корабле
survival_counts = df.group_by("Sex").agg(
    pl.col("Survived").sum())
survival_counts

Sex,Survived
str,i64
"""female""",233
"""male""",109


In [ ]:
# Пассажиры старше 44 лет
df.filter(pl.col("Age") > 44)

PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
i64,i64,i64,str,str,f64,i64,i64,str,f64,str,str
7,0,1,"""McCarthy, Mr. Timothy J""","""male""",54.0,0,0,"""17463""",51.8625,"""E46""","""S"""
12,1,1,"""Bonnell, Miss. Elizabeth""","""female""",58.0,0,0,"""113783""",26.55,"""C103""","""S"""
16,1,2,"""Hewlett, Mrs. (Mary D Kingcome…","""female""",55.0,0,0,"""248706""",16.0,null,"""S"""
34,0,2,"""Wheadon, Mr. Edward H""","""male""",66.0,0,0,"""C.A. 24579""",10.5,null,"""S"""
53,1,1,"""Harper, Mrs. Henry Sleeper (My…","""female""",49.0,1,0,"""PC 17572""",76.7292,"""D33""","""C"""
…,…,…,…,…,…,…,…,…,…,…,…
858,1,1,"""Daly, Mr. Peter Denis ""","""male""",51.0,0,0,"""113055""",26.55,"""E17""","""S"""
863,1,1,"""Swift, Mrs. Frederick Joel (Ma…","""female""",48.0,0,0,"""17466""",25.9292,"""D17""","""S"""
872,1,1,"""Beckwith, Mrs. Richard Leonard…","""female""",47.0,1,1,"""11751""",52.5542,"""D35""","""S"""


# **Ускорение работы с pandas**

In [3]:
import pandas as pd
import bottleneck as bn

In [ ]:
# Загрузка датасета
df = pd.read_csv("train.csv")

In [ ]:
# Подсчет среднего возраста пассажиров и его стандартного отклонения с использованием bottleneck
mean_age = bn.nanmean(df["Age"])
std_age = bn.nanstd(df["Age"])
print(f"Средний возраст пассажиров: {mean_age}")
print(f"Стандартное отклонение возраста: {std_age}")

Средний возраст пассажиров: 29.69911764705882
Стандартное отклонение возраста: 14.516321150817317


In [ ]:
# Умножение значений столбца Fare на 1.3 и сохранение результатов в Fare_new
def multiply_by_1_3(x):
    return x * 1.3

df["Fare_new"] = df["Fare"].apply(multiply_by_1_3)
df["Fare_new"]

,Fare_new
0,9.42500
1,92.66829
2,10.30250
3,69.03000
4,10.46500
...,...
886,16.90000
887,39.00000
888,30.48500
889,39.00000


# **Оптимизация типов pandas**

In [17]:
# Загружаем датасет
df = pd.read_csv('Housing.csv')

In [18]:
# Сохраняем исходное потребление памяти
memory_before = df.memory_usage(deep=True).sum() / (1024 ** 2)  # в МБ
print(f"Потребление памяти до оптимизации: {memory_before:.2f} МБ")

Потребление памяти до оптимизации: 0.25 МБ


In [19]:
# Получение информации об исходном датасете: пропусках и типах данных
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 545 entries, 0 to 544
Data columns (total 13 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   price             545 non-null    int64 
 1   area              545 non-null    int64 
 2   bedrooms          545 non-null    int64 
 3   bathrooms         545 non-null    int64 
 4   stories           545 non-null    int64 
 5   mainroad          545 non-null    object
 6   guestroom         545 non-null    object
 7   basement          545 non-null    object
 8   hotwaterheating   545 non-null    object
 9   airconditioning   545 non-null    object
 10  parking           545 non-null    int64 
 11  prefarea          545 non-null    object
 12  furnishingstatus  545 non-null    object
dtypes: int64(6), object(7)
memory usage: 55.5+ KB


In [20]:
# Анализ используемых значений
for col in df.columns:
    print(f"Столбец {col}: уникальных значений - {df[col].nunique()}, min - {df[col].min()}, max - {df[col].max()}, пример значений - {df[col].unique()[:5]}")


Столбец price: уникальных значений - 219, min - 1750000, max - 13300000, пример значений - [13300000 12250000 12215000 11410000 10850000]
Столбец area: уникальных значений - 284, min - 1650, max - 16200, пример значений - [7420 8960 9960 7500 8580]
Столбец bedrooms: уникальных значений - 6, min - 1, max - 6, пример значений - [4 3 5 2 6]
Столбец bathrooms: уникальных значений - 4, min - 1, max - 4, пример значений - [2 4 1 3]
Столбец stories: уникальных значений - 4, min - 1, max - 4, пример значений - [3 4 2 1]
Столбец mainroad: уникальных значений - 2, min - no, max - yes, пример значений - ['yes' 'no']
Столбец guestroom: уникальных значений - 2, min - no, max - yes, пример значений - ['no' 'yes']
Столбец basement: уникальных значений - 2, min - no, max - yes, пример значений - ['no' 'yes']
Столбец hotwaterheating: уникальных значений - 2, min - no, max - yes, пример значений - ['no' 'yes']
Столбец airconditioning: уникальных значений - 2, min - no, max - yes, пример значений - ['yes

Столбец price - исходный тип данных int64. Данный столбец содержит положительные числовые значения в диапазоне от 1750000 до 13300000. Описанный диапазон значений помещается в int32.
Столбец area - исходный тип данных int64. Данный столбец содержит положительные числовые значения в диапазоне от 1650 до 16200. Описанный диапазон значений помещается в int16.
Столбец bedrooms - исходный тип данных int64. Данный столбец содержит положительные числовые значения в диапазоне от 1 до 6. Описанный диапазон значений помещается в int8.
Столбец bathrooms - исходный тип данных int64. Данный столбец содержит положительные числовые значения в диапазоне от 1 до 4. Описанный диапазон значений помещается в int8.
Столбец stories - исходный тип данных int64. Данный столбец содержит положительные числовые значения в диапазоне от 1 до 4. Описанный диапазон значений помещается в int8.
Столбец parking - исходный тип данных int64. Данный столбец содержит неотрицательные числовые значения в диапазоне от 0 до 3. Описанный диапазон значений помещается в int8.
Столбцы mainroad, guestroom, basement, hotwaterheating, airconditioning, prefarea  - исходный тип данных object. Данные столбцы содержат строки 'yes', 'no'. Это категориальные признаки, которые могут быть представлены булевой переменной: 'yes' - True, 'no' - False.
Столбец furnishingstatus - исходный тип данных object. Данный стобец содержит строки 'furnished', 'semi-furnished', 'unfurnished'. Это категориальные признаки, которые могут быть представлены типом данных category.



In [21]:
binary_cols = ['mainroad', 'guestroom', 'basement', 'hotwaterheating', 'airconditioning', 'prefarea']
for col in binary_cols:
    df[col] = df[col].replace({'yes': True, 'no': False}).astype('bool')

<ipython-input-21-2526c43f0879>:3: FutureWarning: Downcasting behavior in `replace` is deprecated and will be removed in a future version. To retain the old behavior, explicitly call `result.infer_objects(copy=False)`. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  df[col] = df[col].replace({'yes': True, 'no': False}).astype('bool')
<ipython-input-21-2526c43f0879>:3: FutureWarning: Downcasting behavior in `replace` is deprecated and will be removed in a future version. To retain the old behavior, explicitly call `result.infer_objects(copy=False)`. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  df[col] = df[col].replace({'yes': True, 'no': False}).astype('bool')
<ipython-input-21-2526c43f0879>:3: FutureWarning: Downcasting behavior in `replace` is deprecated and will be removed in a future version. To retain the old behavior, explicitly call `result.infer_objects(copy=False)`. To opt-in to the 

In [22]:
df['price'] = df['price'].astype('int32')
df['area'] = df['area'].astype('int16')
df['bedrooms'] = df['bedrooms'].astype('int8')
df['bathrooms'] = df['bathrooms'].astype('int8')
df['stories'] = df['stories'].astype('int8')
df['parking'] = df['parking'].astype('int8')

In [11]:
df['furnishingstatus'] = df['furnishingstatus'].astype('category')

In [23]:
# Получение информации о преобразованном датасете: пропусках и типах данных
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 545 entries, 0 to 544
Data columns (total 13 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   price             545 non-null    int32 
 1   area              545 non-null    int16 
 2   bedrooms          545 non-null    int8  
 3   bathrooms         545 non-null    int8  
 4   stories           545 non-null    int8  
 5   mainroad          545 non-null    bool  
 6   guestroom         545 non-null    bool  
 7   basement          545 non-null    bool  
 8   hotwaterheating   545 non-null    bool  
 9   airconditioning   545 non-null    bool  
 10  parking           545 non-null    int8  
 11  prefarea          545 non-null    bool  
 12  furnishingstatus  545 non-null    object
dtypes: bool(6), int16(1), int32(1), int8(4), object(1)
memory usage: 12.9+ KB


In [24]:
memory_after = df.memory_usage(deep=True).sum() / (1024 ** 2)  # в МБ
print(f"Потребление памяти после оптимизации: {memory_after:.2f} МБ")

Потребление памяти после оптимизации: 0.04 МБ


In [25]:
memory_savings = (memory_before - memory_after) / memory_before * 100
print(f"Экономия памяти: {memory_savings:.2f}%")

Экономия памяти: 82.03%
